storage/redis/peer_crud.go

```diff
@@ -0,0 +1,79 @@
+package redis
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/chihaya/chihaya/bittorrent"
+	redigo "github.com/garyburd/redigo/redis"
+)
+
+func decodePeerKey(pk string) bittorrent.Peer {
+	return bittorrent.Peer{
+		ID:   bittorrent.PeerIDFromString(string(pk[:20])),
+		Port: binary.BigEndian.Uint16([]byte(pk[20:22])),
+		IP:   net.IP(pk[22:]),
+	}
+}
+
+//Adds an expiry to the set if not refreshed
+func addPeer(s *peerStore, infoHash bittorrent.InfoHash, peerType string,
+	pk serializedPeer) error {
+	Key := fmt.Sprintf("%s%s", peerType+":", infoHash)
+	s.conn.Send("MULTI")
+	s.conn.Send("ZADD", Key, time.Now().Unix(), pk)
+	s.conn.Send("EXPIRE", Key, int(s.peerLifetime.Seconds()))
+	_, err := s.conn.Do("EXEC")
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func removePeers(s *peerStore, infoHash bittorrent.InfoHash, peerType string,
+	pk serializedPeer) error {
+	_, err := s.conn.Do("ZREM",
+		fmt.Sprintf("%s%s", peerType+":", infoHash), pk)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Prunes the existing infohash swarm for any old peers before
+// returning range of valid peers
+func getPeers(s *peerStore, infoHash bittorrent.InfoHash, peerType string,
+	numWant int, peers []bittorrent.Peer, excludePeers bittorrent.Peer) (
+	[]bittorrent.Peer, error) {
+	Key := fmt.Sprintf("%s%s", peerType+":", infoHash)
+	_, err := s.conn.Do("ZREMRANGEBYSCORE", Key,
```

Comment:
The cumulative time complexity should be the same in my opinion. It's O(log(N)+M) here with N being the number of elements and M being the number of elements removed. log(N) is infinitesimal regardless of the size of the sorted set. M will be the same regardless of us doing it every announce or in a cleanup daemon. Moreover, making this part of the announce and not a separate daemon eliminates the need for mutexes and sync between multiple instances of this server. This was my reasoning for this. Please let me know your thoughts behind this.